packages/core/src/telemetry/metrics.ts

```diff
@@ -6,38 +6,305 @@
+    attributes: {} as {
+      function_name: string;
+      success: boolean;
+      decision?: 'accept' | 'reject' | 'modify' | 'auto_accept';
+      tool_type?: 'native' | 'mcp';
+    },
+  },
+  [API_REQUEST_COUNT]: {
+    description: 'Counts API requests, tagged by model and status.',
+    valueType: ValueType.INT,
+    assign: (c: Counter) => (apiRequestCounter = c),
+    attributes: {} as {
+      model: string;
+      status_code?: number | string;
+      error_type?: string;
+    },
+  },
+  [TOKEN_USAGE]: {
+    description: 'Counts the total number of tokens used.',
+    valueType: ValueType.INT,
+    assign: (c: Counter) => (tokenUsageCounter = c),
+    attributes: {} as {
+      model: string;
+      type: 'input' | 'output' | 'thought' | 'cache' | 'tool';
+    },
+  },
+  [SESSION_COUNT]: {
+    description: 'Count of CLI sessions started.',
+    valueType: ValueType.INT,
+    assign: (c: Counter) => (sessionCounter = c),
+    attributes: {} as Record<string, never>,
+  },
+  [FILE_OPERATION_COUNT]: {
+    description: 'Counts file operations (create, read, update).',
+    valueType: ValueType.INT,
+    assign: (c: Counter) => (fileOperationCounter = c),
+    attributes: {} as {
+      operation: FileOperation;
+      lines?: number;
+      mimetype?: string;
+      extension?: string;
+      programming_language?: string;
+    },
+  },
+  [INVALID_CHUNK_COUNT]: {
+    description: 'Counts invalid chunks received from a stream.',
+    valueType: ValueType.INT,
+    assign: (c: Counter) => (invalidChunkCounter = c),
+    attributes: {} as Record<string, never>,
+  },
+  [CONTENT_RETRY_COUNT]: {
+    description: 'Counts retries due to content errors (e.g., empty stream).',
+    valueType: ValueType.INT,
+    assign: (c: Counter) => (contentRetryCounter = c),
+    attributes: {} as Record<string, never>,
+  },
+  [CONTENT_RETRY_FAILURE_COUNT]: {
+    description: 'Counts occurrences of all content retries failing.',
+    valueType: ValueType.INT,
+    assign: (c: Counter) => (contentRetryFailureCounter = c),
+    attributes: {} as Record<string, never>,
+  },
+  [MODEL_ROUTING_FAILURE_COUNT]: {
+    description: 'Counts model routing failures.',
+    valueType: ValueType.INT,
+    assign: (c: Counter) => (modelRoutingFailureCounter = c),
+    attributes: {} as {
+      'routing.decision_source': string;
+      'routing.error_message': string;
+    },
+  },
+  [MODEL_SLASH_COMMAND_CALL_COUNT]: {
+    description: 'Counts model slash command calls.',
+    valueType: ValueType.INT,
+    assign: (c: Counter) => (modelSlashCommandCallCounter = c),
+    attributes: {} as {
+      'slash_command.model.model_name': string;
+    },
+  },
+  [EVENT_CHAT_COMPRESSION]: {
+    description: 'Counts chat compression events.',
+    valueType: ValueType.INT,
+    assign: (c: Counter) => (chatCompressionCounter = c),
+    attributes: {} as {
+      tokens_before: number;
+      tokens_after: number;
+    },
+  },
+} as const;
```

Comment:
`as const` is a good start, but there should/must be a type here that can assert this shape.

You can have the best of `as const` and type safety by doing

`as const satisfied Record<SomeEnum, SomeInterface>`

Ideally this type would be `Record<SomeEnum, SomeInterface>`